functions/calcPiTheta.txt

```diff
@@ -0,0 +1,60 @@
+// Written by Nick Bailey, submitted on 2024-06-13
+// Currently affiliated with CNRS and University of Lyon 1
+// This code is provided in the public domain free of restriction, please credit SLiM-Extras (https://github.com/MesserLab/SLiM-Extras) and myself, Nick Bailey, if you use it
+
+// calculate population genetic statistic pi (designated PiTheta to differentiate from mathematical constant pi)
+// calculations originally in Nei and Li 1979 and Tajima 1983
+function (float)calcPiTheta(object<Genome> genomes, [No<Mutation> muts = NULL], [Ni$ start = NULL], [Ni$ end = NULL])
+// conduct various checks for proper usage copied straight from calcWattersonTheta()
+{ if (genomes.length() == 0)
+		stop("ERROR (calcPiTheta()): genomes must be non-empty.");
+	if (community.allSpecies.length() > 1)
+	{
+		species = unique(genomes.individual.subpopulation.species, preserveOrder=F);
+		if (species.length() != 1)
+			stop("ERROR (calcPiTheta()): genomes must all belong to the same species.");
+		if (!isNULL(muts))
+			if (!all(muts.mutationType.species == species))
+				stop("ERROR (calcPiTheta()): muts must all belong to the same species as genomes.");
+	}
+	else
+	{
+		species = community.allSpecies;
+	}
+	
+	if (isNULL(muts))
+		muts = species.mutations;
+	
+	// handle windowing
+	if (!isNULL(start) & !isNULL(end))
+	{
+		if (start > end)
+			stop("ERROR (calcPiTheta()): start must be less than or equal to end.");
+		mpos = muts.position;
+		muts = muts[(mpos >= start) & (mpos <= end)];
+		length = end - start + 1;
+	}
+	else if (!isNULL(start) | !isNULL(end))
+	{
+		stop("ERROR (calcPiTheta()): start and end must both be NULL or both be non-NULL.");
+	}
+
+	// narrow down to the mutations that are actually present in the genomes and aren't fixed
+	p = genomes.mutationFrequenciesInGenomes(muts);
+	muts = muts[(p != 0.0) & (p != 1.0)];
+	
+	// do the calculation
+	// multiply vector of mutation frequencies by integer count of sequences to get vector of counts of variant sequences
+	varCount = genomes.mutationFrequenciesInGenomes(species.mutations) * size(genomes);
+	// total count of sequences subtracted by count of variant sequences equals count of invariant sequences
+	invarCount = size(genomes) - varCount;
+	// count of pairwise differences per site (assuming a biallelic site) is the product of counts of both alleles, this is then summed for all sites 
```

Comment:
this phrasing in the comment is confusing, since sites are not necessarily biallelic

Reply:
Agreed though not sure my new wording is any better. I mention this a bit now in my bigger comment and within the docstrings. I also changed the comment within the script now so it states where I got the idea of just multiplying ref and var alleles, since I don't see that notation often.